source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <thread>
+#include <type_traits>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/singleton/instance.h"
+
+#include "common/common/assert.h"
+#include "common/common/non_copyable.h"
+#include "common/common/thread_synchronizer.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace SharedPool {
+
+/**
+ * Used to share objects that have the same content.
+ * control the life cycle of shared objects by reference counting
+ *
+ * Note:  ObjectSharedPool needs to be created in the main thread,
+ * all the member methods can only be called in the main thread,
+ * it does not have the ownership of object stored internally, the internal storage is weak_ptr,
+ * when the internal storage object destructor executes the custom deleter to remove its own
+ * weak_ptr from the ObjectSharedPool.
+ *
+ * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
+ * main thread, or that ObjectSharedPool destruct before the program exit
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+
+  void deleteObject(const size_t hash_key) {
+    if (std::this_thread::get_id() == thread_id_) {
+      // There may be new inserts with the same hash value before deleting the old element,
+      // so there is no need to delete it at this time.
+      if (object_pool_.find(hash_key) != object_pool_.end() &&
+          object_pool_[hash_key].use_count() == 0) {
+        object_pool_.erase(hash_key);
+      }
+    } else {
+      // Most of the time, the object's destructor occurs in the main thread, but with some
+      // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
+      // safe, the deleteObject needs to be delivered to the main thread.
```

Comment:
What are these exceptions @zyfjeff? It seems pretty dangerous to allow objects created on one thread (main) be destructed on another (worker).

Reply:
https://github.com/envoyproxy/envoy/pull/10042#discussion_r379155904
 Some scenarios are mentioned in this comments, and I'm here to ensure thread safety by `dispatcher::post` converting these scenarios that are destructed in work threads into main thread.

Reply:
I see; I think it would be helpful to have that example (of EDS cleanup on the main thread) in this comment block, otherwise the reasoning gets lost. Seems reasonable to destruct on the main thread as you do here.